replication/AmbryReplication.tla

```diff
@@ -0,0 +1,328 @@
+---------------------- MODULE AmbryReplication ---------------------
+
+EXTENDS Naturals
+(***************************************************************************)
+(* This specification describes the replication in Ambry, in which         *)
+(* messages can come to replicas of an Ambry partition. The messages can   *)
+(* be "PUT", "DELETE", "TTL UPDATE" (undelete will be covered later).      *)
+(***************************************************************************)
+
+
+(* KNOWN ISSUES IN THE MODEL: (fixes are WIP) *)
+(* The sucess target of put, delete and update-ttl operations is not configurable *)
+(* Replication replicates everything in each iteration. It should rather replicate only those blobs that need replication *)
```

Comment:
Does this mean that in a state change from n -> n' all records are replicated at once?  Or do they trickle in at each state?

Reply:
Lets define two things
1. "local store state" of a replica :::  this is the set of records that a replica has (i.e, this is the entire set of PUT/UPDATE_TTL/DELETE messages that a replica has seen and has applied locally).
2. What does the state change n->n' in this context :::  The state change n->n' means that replica A replicates records to its own "local store" from replica B (ambry's replication is pull-based).


So what the issue means is that in the n->n' state change, replica B will give its entire "local-store-state" to replica A. Replica A can then decide which of the B's "local-store-state" it already has and which ones are new. Replica A will then update the new records in its "local store".

To more accurately model Ambry's replication logic, what should happen instead, is that replica A should remember (i.e, maintain a ReplicaToken) how much it has already caught up against replica B. Then in the n->n' state change, replica A will only ask for B's "local storage state" after the ReplicaToken (previously remembered point), and hence B will only give its new records to replica A.

Reply:
Got it.  So I'm inferring this token approach is what you have in mind for "should ... replicate only those blobs that need replication"?

Reply:
Yes, thats right.
I am also trying to make up my mind about whether the additional complexity it brings will really help with catching bugs. The extra complexity will come from
1. ReplicaToken data structure == Each replica will need to track a ReplicaToken for all its peers, so the data structure will be more like a function of records where each record itself is a function of records
2. To maintain ReplicaToken, the "local store" being represented by the "storedMessagesPerReplica" variable will need to change from the function "Replica -> Set of Records" to "Replica -> An ordered list of records".

If we don't want to model a case where a replica's local store might get wiped out totally then the additional complexity might not be worth it. But I think that case is a legit case that should be modeled.